### ISSUE TITLE
What processes in LIDAR-RCNN are specific for waymo dataset?

### ISSUE THREAD

**Issue by QingXIA233** [TuSimple/LiDAR_RCNN#40]
Hello, just like the title saying, I wonder what are the specific processes for WOD, which means if I want to use LIDAR R-CNN on my own dataset, I have to do it differently. I already change the **data_processor** and everything I can think of in the **loader** and **creat_results** that are respect to waymo dataset, then I use the refined results to perform evaluation on my own dataset. However, I got NAN on rotation error, and the MAP is pretty low. 
<img width="298" alt="issue2" src="https://user-images.githubusercontent.com/64100948/203033403-202c9527-1a22-468a-9c71-c9870c9396a1.png">

Therefore, I'm confused about some subtle processes that are performed just for waymo not for other datasets. For example, compute heading residual is necessary for using LiDAR R-CNN? Did you guys use rotation in some sublte ways? （In my dataset, the rotation is according to y axis, while in your code, it's x axis, but the way of computing rotZ is the same, I already changed it.）
![image](https://user-images.githubusercontent.com/64100948/203032516-76ef7e66-047e-46da-ad3b-5d96e3083034.png)

This bug has been driving me crazy, that's why my issue description above is a bit messy, forgive me please. I would be grateful if you could provide me some hints. Thank you a lot. Save this almost desperate kid, please.🥺

**Comment by Lzc6996** [comment]
@QingXIA233 
I am sorry for the suffering. This repo is build for the official code for [our paper](https://arxiv.org/abs/2103.15297). So it is unfortunately all build for WOD.  At least you need to adapt [loader](https://github.com/tusen-ai/LiDAR_RCNN/blob/master/src/LiDAR_RCNN/datasets/waymo/loader.py), [data_utils](https://github.com/tusen-ai/LiDAR_RCNN/blob/master/src/LiDAR_RCNN/datasets/waymo/data_utils.py), [data_processor](https://github.com/tusen-ai/LiDAR_RCNN/tree/master/tools/data_processer), [config ](https://github.com/tusen-ai/LiDAR_RCNN/tree/master/config)in your dataset. 
The key difference may come from the processing of coordinate system.  And the[ IOU threshold](https://github.com/tusen-ai/LiDAR_RCNN/blob/c693828a5ff2d65ab7293c84837df9e3f6a7786b/config/lidar_rcnn.yaml#L35) in config may should be noticed.

**Comment by QingXIA233** [comment]
Thank you for the help. Yes, I already changed the files you mentioned above. There are several questions about the code that are confusing me:
1. About the extract_points, is the heading parameter of the pred_bbox_bev important here?
<img width="500" alt="issue3" src="https://user-images.githubusercontent.com/64100948/203472999-2c56d503-d663-487a-b32f-95b5104fc0c9.png">
Because, in WOD, the rotation is around z axis and starting from pos x axis in my understanding, but in my own dataset, the rotation is around z axis and starting form -y axis. So, if heading is used in extrac_points, should I change it according to my own dataset setting? (The heading of my dataset bboxes is shown in the figure below)
<img width="500" alt="issue4" src="https://user-images.githubusercontent.com/64100948/203473011-64cf9608-cbd5-45bc-bbd5-348a1f4ada92.png">

2. In loader.py, when moving gt box to pred center, why use minus here, shouldn't it just be = instead of -= ?
<img width="500" alt="issue5" src="https://user-images.githubusercontent.com/64100948/203473455-0add68fb-ef71-49f5-bded-68fc2ef69ca6.png">

**Comment by Lzc6996** [comment]
@QingXIA233 
1. Yes, 6 from pred_lst is heading.  You should also change the extract_points.
2. this also because the definition of heading. If your dataset is different, please do some visualization to check it is correct.

**Comment by QingXIA233** [comment]
Yes, I changed the extrac_points. During the process of moving gt box to pred center, I understand the first line: gt_box[:3] -= proposal[:3}, but why the rotation angle is -propasal heading? In your paper, I see this:
![image](https://user-images.githubusercontent.com/64100948/203714893-99a69c58-4e1e-4cfc-8c0e-f24adfe74c9c.png)
So, shouldn't the rotation angle be the difference between the gt heading and proposal heading?
BTW, I wonder where I can find the definition of the WOD axes definition (x, y, z and yaw ratation). For example, is the z-axis vertical to the car or the direction of car going?

**Comment by QingXIA233** [comment]
Hi, thank you for the warm patience. I just got confused by the transformations in the code:
here, when transfer pcd to pred box coords, you use the rotz matrix of proposal heading
<img width="500" alt="issue7" src="https://user-images.githubusercontent.com/64100948/203739666-01081074-9a2d-4f69-bc26-8ef67babae94.png">

However, when transfer gt to pred box coords, you use the rotz matrix of  negative proposal heading
![image](https://user-images.githubusercontent.com/64100948/203739552-47226df3-2cab-43a1-9174-71086ff04110.png)

I didn't get the logic here, could you please enlighten me a little? Thanks

**Comment by Lzc6996** [comment]
@QingXIA233 
This is the difference between left and right multiplication.

**Comment by xiuzhizheng** [comment]
@QingXIA233 
Hi, what performace of lidar rcnn on your own dataset?

**Comment by QingXIA233** [comment]
> @QingXIA233 Hi, what performace of lidar rcnn on your own dataset?

Hi, actually, the performance on my own dataset is surprisingly terrible. I have 11 classes, and when I tried to refine the first 4 classes, all of them got around 10 points declination. There is another issue which is about performance declination on nuscenes dataset. The author mentioned about tuning the ratio of the positive/negative. I tuned the IOU threshold and still the refinement is not acheived.
